src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,230 @@
+    exact upper_half_plane.denom_ne_zero A x, },
+  rw [h1, h22],
+  ring,
+end
+
+lemma slash_mul_SL2 (k1 k2 : ℤ) (A : SL(2, ℤ)) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1 + k2, A] = (f ∣[k1, A]) * (g ∣[k2, A]) :=
+calc (f * g) ∣[k1 + k2, (A : GL(2, ℝ)⁺)] = _ • (f ∣[k1, A]) * (g ∣[k2, A]) : slash_mul _ _ _ _ _
+... = (1:ℝ) • (f ∣[k1, A]) * (g ∣[k2, A]) : by simp [-matrix.special_linear_group.coe_matrix_coe]
+... = (f ∣[k1, A]) * (g ∣[k2, A]) : by simp
+
+lemma slash_mul_subgroup (k1 k2 : ℤ) (Γ : subgroup SL(2, ℤ)) (A : Γ) (f g : ℍ → ℂ) :
+  (f * g) ∣[k1 + k2, A] = (f ∣[k1, A]) * (g ∣[k2, A]) :=
+begin
+  have : (((↑ₘ(A : GL(2,ℝ)⁺)).det) : ℝ) = 1,
+  by { simp only [coe_coe,matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.det_coe], },
+  have t1 := slash_mul k1 k2 A f g,
+  rw this at t1,
+  simp only [coe_coe, one_smul] at t1,
+  convert t1,
+end
+
+/-- A function `f : ℍ → ℂ` is weakly modular, of weight `k ∈ ℤ` and level `Γ`, if for every matrix .
+ `γ ∈ Γ` we have `f(γ • z)= (c*z+d)^k f(z)` where `γ= ![![a, b], ![c, d]]`, and it acts on `ℍ`
+  via Möbius transformations. -/
+lemma weakly_modular_mem' (k : ℤ) (Γ : subgroup SL(2, ℤ)) (f : ℍ → ℂ) :
+  f ∈ weakly_modular_form k Γ ↔ ∀ γ : Γ, ∀ z : ℍ,
+  f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ))^k * f z :=
+begin
+  simp only [weakly_modular_mem, function.funext_iff, slash_action_eq_slash, slash],
+  refine forall₂_congr _,
+  intros γ z,
+  convert inv_mul_eq_iff_eq_mul₀ _ using 2,
+  { rw mul_comm,
+    simp [-matrix.special_linear_group.coe_matrix_coe] },
+  { exact zpow_ne_zero _ (denom_ne_zero _ _) },
+end
+
+lemma mul_modular (k_1 k_2 : ℤ) (Γ : subgroup SL(2, ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ weakly_modular_form k_1 Γ) (hg : g ∈ weakly_modular_form k_2 Γ) :
+  f * g ∈ weakly_modular_form (k_1 + k_2) Γ :=
+begin
+  simp only [weakly_modular_mem', pi.mul_apply, coe_coe] at *,
+  intros γ z,
+  rw [(hf γ z), (hg γ z)],
+  have pown := zpow_add₀ (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z) k_1 k_2,
+  simp only [upper_half_plane.denom, coe_fn_coe_base, ne.def,
+    matrix.general_linear_group.coe_fn_eq_coe, coe_coe] at pown,
+  rw pown,
+  ring,
+end
+
+/-- A function `f : ℍ → ℂ` is a modular form weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular and bounded at infinity. -/
+structure is_modular_form (k : ℤ) (Γ : subgroup SL(2, ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_form k Γ)
+(infinity : ∀ (A : SL(2, ℤ)), is_bounded_at_im_infty (f ∣[k, A]))
+
+/-- A function `f : ℍ → ℂ` is a cusp form of weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular, and zero at infinity. -/
+structure is_cusp_form (k : ℤ) (Γ : subgroup SL(2, ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_form k Γ)
+(infinity : ∀ (A : SL(2, ℤ)), is_zero_at_im_infty (f ∣[k, A]))
+
+lemma is_modular_form_of_is_cusp_form {f : ℍ → ℂ}
```

Comment:
It would be nice to enable projection notation here: `is_cusp_form.is_modular_form`.